supabase/migrations/30_manual_billing.sql

```diff
@@ -0,0 +1,354 @@
+  ),
+  -- The last day captures the cumulative billed period.
+  last_day as (
+    select * from daily_line_items
+    order by ts desc limit 1
+  ),
+  -- If we're reporting for the whole tenant then gather billing adjustment line-items.
+  adjustments as (
+    select coalesce(jsonb_agg(
+      jsonb_build_object(
+        'description', detail,
+        'count', 1,
+        'rate', usd_cents,
+        'subtotal', usd_cents
+      )
+    ), '[]') as adjustment_line_items
+    from internal.billing_adjustments a
+    where a.tenant = billed_prefix and a.billed_month = billing_report_202308.billed_month
+  )
+  select into
+    -- Block of variables being selected into.
+    o_daily_usage,
+    o_data_gb,
+    o_line_items,
+    o_recurring_fee,
+    o_task_hours,
+    o_trial_credit,
+    o_trial_start,
+    o_trial_range,
+    o_billed_range
+    -- The actual selected columns.
+    daily_usage,
+    data_gb,
+    data_line_items || task_line_items || adjustment_line_items,
+    recurring_fee,
+    task_hours,
+    trial_credit,
+    trial_start,
+    trial_range,
+    billed_range
+  from daily_array_and_trial_credit, last_day, adjustments, vars;
+
+  -- Add line items for recurring service fee & free trial credit.
+  if o_recurring_fee != 0 then
+    o_line_items = jsonb_build_object(
+      'description', 'Recurring service charge',
+      'count', 1,
+      'rate', o_recurring_fee,
+      'subtotal', o_recurring_fee
+    ) || o_line_items;
+  end if;
+
+  -- Display a (possibly zero) free trial credit if the trial range overlaps the billed range
+  if o_trial_range && o_billed_range then
+    o_line_items = o_line_items || jsonb_build_object(
+      'description', format('Free trial credit (%s - %s)', lower(o_trial_range), (upper(o_trial_range) - interval '1 day')::date),
+      'count', 1,
+      'rate', -o_trial_credit,
+      'subtotal', -o_trial_credit
+    );
+  end if;
+
+  -- Roll up the final subtotal.
+  select into o_subtotal sum((l->>'subtotal')::numeric)
+    from jsonb_array_elements(o_line_items) l;
+
+  return jsonb_build_object(
+    'billed_month', billed_month,
+    'billed_prefix', billed_prefix,
+    'daily_usage', o_daily_usage,
+    'line_items', o_line_items,
+    'processed_data_gb', o_data_gb,
+    'recurring_fee', o_recurring_fee,
+    'subtotal', o_subtotal,
+    'task_usage_hours', o_task_hours,
+    'trial_credit', coalesce(o_trial_credit, 0),
+    'trial_start', o_trial_start
+  );
+
+end
+$$ language plpgsql volatile security definer;
+
+-- Note: have to redefine this to know about internal.billing_report
+-- Calculate the specified month's billing report for every tenant
+-- and save those reports to billing_historicals.
+create or replace function internal.freeze_billing_month(billed_month timestamptz)
+returns integer as $$
+declare
+    tenant_row record;
+    tenant_count integer = 0;
+begin
+    for tenant_row in select tenant as tenant_name from tenants loop
+        insert into billing_historicals
```

Comment:
move `billing_historicals` into `internal` schema too? and we can drop RLS etc and revoke query access for `authenticated`.